source/opt/loop_descriptor.cpp

```diff
@@ -153,5 +250,122 @@ void LoopDescriptor::PopulateList(const Function* f) {
   }
 }
 
+ir::BasicBlock* Loop::FindConditionBlock(const ir::Function& function) {
+  ir::BasicBlock* condition_block = nullptr;
+
+  opt::DominatorAnalysis* dom_analysis =
+      ir_context_->GetDominatorAnalysis(&function, *ir_context_->cfg());
+  ir::BasicBlock* bb = dom_analysis->ImmediateDominator(loop_merge_);
+
+  if (!bb) return nullptr;
+
+  if (bb->ctail()->opcode() == SpvOpBranchConditional) {
+    condition_block = bb;
+  }
+
+  return condition_block;
+}
+
+bool Loop::FindNumberOfIterations(const ir::Instruction* induction,
+                                  const ir::Instruction* branch_inst,
+                                  size_t* iterations) const {
+  // From the branch instruction find the branch condition.
+  opt::analysis::DefUseManager* def_use_manager =
+      ir_context_->get_def_use_mgr();
+
+  // Condition instruction from the OpConditionalBranch.
+  ir::Instruction* condition =
+      def_use_manager->GetDef(branch_inst->GetSingleWordOperand(0));
+
+  assert(condition->opcode() == SpvOpSLessThan);
+
+  // The right hand side operand of the operation.
+  const ir::Instruction* rhs_inst =
+      def_use_manager->GetDef(condition->GetSingleWordOperand(3));
+
+  // Exit out if we couldn't resolve the rhs to be a constant integer.
+  uint32_t condition_value = 0;
+  if (!GetConstant(rhs_inst, ir_context_, &condition_value)) return false;
```

Comment:
I think you could use the constant manager here.

```
const Constant* upper_bound = const_mgr->FindDeclaredConstant(condition->GetSingleWordOperand(3));
if (!upper_bound) return false;
// must be integer because of the opcode on the condition.
uint32_t condition_value = upper_bound->AsIntegerConstant()->GetUInt32BitValue(); 
```

If we can get rid of GetConstant, we can reduce the amount of code.